haskell-api.md

```diff
@@ -0,0 +1,236 @@
+than our multi-id path-pieces of today.
+
+---
+
+## Response fields
+
+- Always include `id`s
+- Within reason, choose attributes directly for the work that is motivating the
+  endpoint
+- Include related resources in abbreviated form when Frontend needs extra data
+  about them. For example:
+  - `students: [{id:, firstName:, lastName:}]`
+  - `school: {id:, name:}`
+
+## Status Codes
+
+- 201 for creation
+- 204 for no content (i.e. our `Empty` responses)
+- 400s MUST follow our `ValidationError` machinery
+
+## Pagination
+
+- All list-returning routes of modest size should be paginated via `Yesod.Page`
+
+## Modules Structure
+
+- Fully expand all parents to start
+
+  This means we won't break Frontend when the tree expands over time
+
+- Fully name-space all parents and routes (see example)
+
+  **Caveats**:
+
+  - There's no need to `V3`-prefix everything
+  - Things can "start over" when there's the "single below a list" situation,
+    e.g. `TeacherP x $ TeacherR` instead of `TeachersP $ TeachersTeacherP x $
+    TeachersTeacherR`, which is a bit buffalo-buffalo
+  - The "start over" caveat does not apply to path naming (i.e. the above
+    example will still have the module `Teachers/Teacher.hs`)
+
+- Name modules to directly match parents and routes, which implies one Handler
+  module per route (see example)
+
+  **Caveats**:
+
+  - In reality, we need to prevent collisions between versions through
+    additional prefixing of route constructors. We ignore this in this guide.
+
+Example:
+
+```
+/3 V3P:                         --> Handlers/V3
+  /students StudentsP:          --> Handlers/V3/Students
+    / StudentsR GET             --> Handlers/V3/Students.hs
+    /#StudentId StudentP:       --> Handlers/V3/Students/Student
+      / StudentR GET            --> Handlers/V3/Students/Student.hs
+      /courses StudentCoursesP: --> Handlers/V3/Students/Student/Courses
+        / StudentCoursesR GET   --> Handlers/V3/Students/Student/Courses.hs
+```
+
+## Module Structure
+
+- Prefer in-module request/response types
+
+  - Request type naming: `{Resource}{Method}Body`
+  - Response type naming: `{Resource}{Method}`
+  - NOTE: list responses are `[{Resource}{Method}]`, not `{Resources}{Method}`
+
+  ```hs
+  -- Good
+  data TeacherGet
+
+  getTeachersR :: Handler [TeacherGet]
+  getTeachersR = undefined
+
+  data TeacherPost
+  data TeacherPostBody
+
+  postTeacherR :: Handler TeacherPost
+  postTeacherR = do
+    body <- requireJsonBody @TeacherPostBody
+
+  -- Bad
+  data Teachers
+
+  getTeachersR :: Handler Teachers
+  getTeachersR = undefined
+
+  data PostTeacher
+
+  postTeacherR :: Handler (Entity Teacher)
+  postTeacherR = do
+    body <- requireJsonBody @PostTeacher
+  ```
+
+- Prefer in-module data-access functions, until they require sharing
+- Order your module as:
+  - Request/response type(s)
+  - Handler that uses it
+  - Repeat if more than one handler
+  - Internal functions
+- Document your endpoints
```

Comment:
:+1: Thank you for adding this to the guide, documentation on the v2 endpoints is pretty holey.